README.md
pyramid_scheduler
=================

A pyramid plugin that allows asynchronous and deferred task scheduling and management.

CHANGELOG.rst
=========
ChangeLog
=========


v0.3.4
======

* Added support for alternate specification of job execution via
  `scheduler.delegator` config


v0.3.3
======

* Improved "async" jobs to run immediately instead of relying on
  APScheduler (thus improving time-to-execution)


v0.3.2
======

* Removed `distribute` dependency


v0.3.1
======

* Fix for Python 3 support
  (thanks to github.com/farvour -- issue #7)


v0.3.0
======

* Added Python 3 support
  (thanks to github.com/farvour -- issue #3)
* Implemented workaround for kombu (python3 + pickle + bzip2) issue


v0.2.5
======

* Added ultra-basic unit test
* Fixed serializer to work with kombu 3.0.8
  (thanks to github.com/mazvv -- issue #2)


v0.2.4
======

* First tagged release


pyramid_scheduler/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/12/18
# copy: (C) Copyright 2013-EOT Cadit Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import os
import unittest
import tempfile
import shutil
import sqlalchemy as sa
import json
import uuid
import webtest
import time
import threading
import kombu.transport.sqlalchemy.models as ksam
from kombu.transport.sqlalchemy.models import metadata, Queue, Message
from pyramid.config import Configurator
from pyramid.response import Response

#------------------------------------------------------------------------------
storedData = []
storedDataCond = threading.Condition()
def store_some_data(**data):
  with storedDataCond:
    storedData.append(data)
    storedDataCond.notify()

#------------------------------------------------------------------------------
def ReflectorApp(settings={}):
  engine = sa.engine_from_config(settings, 'sqlalchemy.')
  config = Configurator(settings=settings)
  def reflect_params(request):
    return Response(json.dumps(dict(request.params)))
  def deferred_store_params(request):
    request.registry.scheduler.add_date_job(
      store_some_data, time.time() + 0.5, kwargs=dict(request.params))
    return Response('ok')
  config.add_route('reflect', '/reflect/*path')
  config.add_view(reflect_params, route_name='reflect')
  config.add_route('deferred', '/store/*path')
  config.add_view(deferred_store_params, route_name='deferred')
  return config.make_wsgi_app()

#------------------------------------------------------------------------------
class TestScheduler(unittest.TestCase):

  #----------------------------------------------------------------------------
  @classmethod
  def setUpClass(klass):
    # create a master database
    fp = tempfile.NamedTemporaryFile(
      suffix='.db', prefix='pyramid_scheduler-unittest-master-', delete=False)
    klass.masterdb_path = fp.name
    fp.close()
    engine = sa.create_engine('sqlite:///' + klass.masterdb_path)
    # kombu 3.0.8 compat
    if hasattr(ksam, 'class_registry'):
      class Queue(ksam.Queue, ksam.ModelBase):
        __tablename__ = 'kombu_queue'
      class Message(ksam.Message, ksam.ModelBase):
        __tablename__ = 'kombu_message'
    metadata.create_all(engine)
    # todo: any other default things?...
    # TODO: this is a hack. fix! for some reason, when auto-inserting the
    #       queue names, i *sometimes* get the following error:
    #         IntegrityError: (IntegrityError) column name is not unique
    #         u'INSERT INTO kombu_queue (name) VALUES (?)' ('test-queue',)
    engine.execute("INSERT INTO kombu_queue (name) VALUES ('test-queue')")

  #----------------------------------------------------------------------------
  @classmethod
  def tearDownClass(klass):
    os.unlink(klass.masterdb_path)

  #----------------------------------------------------------------------------
  def setUp(self):
    # replicate the master database and create a test application
    fp = tempfile.NamedTemporaryFile(
      suffix='.db', prefix='pyramid_scheduler-unittest-replica-', delete=False)
    self.dbpath = fp.name
    self.dburl  = 'sqlite:///' + self.dbpath
    fp.close()
    shutil.copyfile(self.masterdb_path, self.dbpath)

  #----------------------------------------------------------------------------
  def tearDown(self):
    os.unlink(self.dbpath)

  #----------------------------------------------------------------------------
  def initApp(self, **settings):
    appset = {
      'sqlalchemy.url'        : self.dburl,
      'pyramid.includes'      : 'pyramid_scheduler',
      'scheduler.combined'    : 'true',
      'scheduler.queues'      : 'test-queue',
      'scheduler.broker.url'  : 'kombu.transport.sqlalchemy:Transport+' + self.dburl,
    }
    appset.update(settings or {})
    self.app     = ReflectorApp(settings=appset)
    self.testapp = webtest.TestApp(self.app)

  #----------------------------------------------------------------------------
  def test_testapp(self):
    self.initApp()
    res = self.testapp.get('/reflect/params/?hello=world')
    self.assertEqual(res.body, b'{"hello": "world"}')
    val = str(uuid.uuid4())
    res = self.testapp.get('/reflect/params/?value=' + val)
    self.assertEqual(res.body, b'{"value": "' + val.encode('ascii') + b'"}')

  #----------------------------------------------------------------------------
  def test_kombu308_newSerializerDefault(self):
    self.initApp()
    self.assertEqual(storedData, [])
    val = str(uuid.uuid4())
    res = self.testapp.get('/store/?val=' + val)
    self.assertEqual(res.body, b'ok')
    with storedDataCond:
      if not storedData:
        storedDataCond.wait(timeout=3)
      self.assertEqual(storedData, [{'val': val}])


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


pyramid_scheduler/api.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# lib:  pyramid_scheduler.engine
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/15
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

import sys, time, pickle
from .util import adict

#------------------------------------------------------------------------------
class InvalidJob(Exception): pass
class InvalidQueue(Exception): pass
class InvalidMessage(Exception): pass

#------------------------------------------------------------------------------
class Event(adict):

  MESSAGE       = 'message'
  JOB_CREATED   = 'job.created'
  JOB_EXECUTED  = 'job.executed'
  JOB_CANCELED  = 'job.canceled'
  JOB_REMOVED   = 'job.removed'

  #----------------------------------------------------------------------------
  def __init__(self, type, **params):
    self.type = type
    self.update(params)

  #----------------------------------------------------------------------------
  def __repr__(self):
    return '<%s.Event %s=%r>' \
      % (__name__, self.type,
         {k: v for k, v in self.items() if k != 'type'})

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------


pyramid_scheduler/util.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# lib:  pyramid_scheduler.util
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/15
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

import time, datetime, re, uuid
import apscheduler.events

#------------------------------------------------------------------------------
class InvalidDuration(Exception): pass

#------------------------------------------------------------------------------
class adict(dict):
  def __getattr__(self, key):
    if key.startswith('__') and key.endswith('__'):
      # note: allows an adict to be pickled with protocols 0, 1, and 2
      #       which treat the following specially:
      #         __getstate__, __setstate__, __slots__, __getnewargs__
      return dict.__getattr__(self, key)
    return self.get(key, None)
  def __setattr__(self, key, value):
    self[key] = value
    return self
  def __delattr__(self, key):
    if key in self:
      del self[key]
    return self
  def update(self, *args, **kw):
    args = [e for e in args if e]
    dict.update(self, *args, **kw)
    return self
  @staticmethod
  def __dict2adict__(subject, recursive=False):
    if isinstance(subject, list):
      if not recursive:
        return subject
      return [adict.__dict2adict__(val, True) for val in subject]
    if not isinstance(subject, dict):
      return subject
    ret = adict(subject)
    if not recursive:
      return ret
    for key, val in ret.items():
      ret[key] = adict.__dict2adict__(val, True)
    return ret

#------------------------------------------------------------------------------
durRE   = re.compile(r'\s*(\d+|\d*\.\d+)\s*([wdhms])\s*')
durMult = dict(w=604800, d=86400, h=3600, m=60, s=1)
def asdur(spec):
  ret = 0
  curspec = spec
  while True:
    match = durRE.match(curspec)
    if not match:
      raise InvalidDuration(spec)
    ret += float(match.group(1)) * durMult[match.group(2)]
    curspec = spec[match.end():]
    if len(curspec) <= 0:
      break
  return ret

#------------------------------------------------------------------------------
def addPrefix(src, prefix):
  return adict({prefix + k: v for k, v in src.items()})

#------------------------------------------------------------------------------
def makeID():
  return str(uuid.uuid4())

#------------------------------------------------------------------------------
def apsEventCode2String(code):
  for key in dir(apscheduler.events):
    if code is getattr(apscheduler.events, key):
      return key
  return None

#------------------------------------------------------------------------------
def apsevents_repr(self):
  if isinstance(self, apscheduler.events.JobStoreEvent):
    if not hasattr(self, 'job'):
      return '<apscheduler.events.JobStoreEvent code=%s, store=%s>' \
        % (apsEventCode2String(self.code), self.alias)
    return '<apscheduler.events.JobStoreEvent code=%s, store=%s, job=%r>' \
      % (apsEventCode2String(self.code), self.alias, self.job)
  if isinstance(self, apscheduler.events.JobEvent):
    return '<apscheduler.events.JobEvent code=%s, job=%r>' \
      % (apsEventCode2String(self.code), self.job)
  return '<apscheduler.events.SchedulerEvent code=%s>' \
    % (apsEventCode2String(self.code),)
apscheduler.events.SchedulerEvent.__repr__ = apsevents_repr

#------------------------------------------------------------------------------
def cull(**kw):
  return {key : value for key, value in kw.items() if value is not None}

#------------------------------------------------------------------------------
def resolve(spec):
  if not spec or callable(spec):
    return spec
  if ':' in spec:
    spec, attr = spec.split(':', 1)
    return getattr(resolve(spec), attr)
  spec = spec.split('.')
  used = spec.pop(0)
  found = __import__(used)
  for cur in spec:
    used += '.' + cur
    try:
      found = getattr(found, cur)
    except AttributeError:
      __import__(used)
      found = getattr(found, cur)
  return found

#------------------------------------------------------------------------------
def now():
  return time.time()

#------------------------------------------------------------------------------
def ts2dt(ts):
  # NOTE: APS requires naive datetimes... ugh. so need to return a
  # naive datetime in local timezone.
  return datetime.datetime.fromtimestamp(ts)

#------------------------------------------------------------------------------
def dt2ts(dt):
  # NOTE: APS requires naive datetimes... ugh. so need to ignore the
  # fact that the datetime is naive.
  return float(time.mktime(dt.timetuple()))

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


pyramid_scheduler/broker.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# lib:  pyramid_scheduler.broker
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/15
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

'''
The `pyramid_scheduler.broker` module deals with the communication
between job producers and job executers (consumers). The default
implementation uses `kombu` as the underlying engine.
'''

import time, logging, threading
import kombu, kombu.pools, kombu.common
from kombu.mixins import ConsumerMixin
import six

from . import api
from .util import adict

log = logging.getLogger(__name__)

#------------------------------------------------------------------------------
class KombuWorker(ConsumerMixin):
  def __init__(self, engine, connection):
    self.engine     = engine
    self.connection = connection
  def get_consumers(self, Consumer, channel):
    return [Consumer(
      queues    = self.engine.queues,
      callbacks = [self.event],
      accept    = ['pickle'],
    )]
  def event(self, event, message):
    log.debug('received message: %r', event)
    self.engine.scheduler._handleEvent(event)
    message.ack()

#------------------------------------------------------------------------------
class Engine(object):

  #----------------------------------------------------------------------------
  def __init__(self, scheduler, conf):
    self.scheduler  = scheduler
    self.thread     = None
    self.exchange   = kombu.Exchange('pyramid_scheduler', type='topic')
    self.queues     = [
      kombu.Queue(queue, exchange=self.exchange, routing_key=queue)
      for queue in scheduler.conf.queues]
    self.connection = kombu.Connection(conf.get('broker.url'))
    self.serializer = conf.get('broker.serializer', 'pickle')
    # note: disabling compression for PY3 since kombu 2.5.10 fails
    #       with pickle+bzip2...
    # todo: remove this workaround when kombu is fixed...
    self.compressor = conf.get('broker.compressor', None if six.PY3 else 'bzip2')

  #----------------------------------------------------------------------------
  def send(self, message, queue=None):
    if isinstance(message, six.string_types):
      message = api.Event('message', message=message)
    if not isinstance(message, api.Event):
      raise api.InvalidMessage('broker messages must be instances of'
                               ' pyramid_scheduler.api.Event, not %s: %r' %
                               (type(message), message,))
    self._send(queue, message)

  #----------------------------------------------------------------------------
  def startConsumer(self, queues):
    if queues:
      log.warning('queue selection list is specified but not supported yet -- ignoring')
    # TODO: implement use of `queues`...
    self.thread = threading.Thread(target=self._run)
    self.thread.daemon = True
    self.thread.start()

  #----------------------------------------------------------------------------
  def _run(self):
    while True:
      try:
        log.debug('acquiring event consumer')
        with kombu.pools.connections[self.connection].acquire(block=True) as conn:
          try:
            KombuWorker(self, conn).run()
          except Exception:
            log.exception('scheduler event consumer failed during connection handling')
      except Exception:
        log.exception('scheduler event consumer failed to acquire/listen to a connection')
        time.sleep(5)

  #----------------------------------------------------------------------------
  def _send(self, queue, payload):
    if not queue:
      for q in [q.name for q in self.queues]:
        self._send(q, payload)
      return
    if queue not in [q.name for q in self.queues]:
      raise api.InvalidQueue(queue)
    log.debug('sending message %r to %s', payload, queue)
    with kombu.pools.producers[self.connection].acquire(block=True) as producer:
      kombu.common.maybe_declare(self.exchange, producer.channel)
      producer.publish(
        payload,
        serializer   = self.serializer,
        compression  = self.compressor,
        exchange     = self.exchange,
        routing_key  = queue,
        declare      = self.queues,
      )

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


pyramid_scheduler/scheduler.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# lib:  pyramid_scheduler.scheduler
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/04/15
# copy: (C) Copyright 2013 Cadit Inc., see LICENSE.txt
#------------------------------------------------------------------------------

'''
The pyramid-scheduler scheduling service.
'''

import time
import datetime
import logging
import threading
from contextlib import contextmanager

import transaction
import apscheduler
import apscheduler.scheduler
import apscheduler.events
from pyramid.settings import asbool, aslist
from apscheduler.jobstores.ram_store import RAMJobStore
from apscheduler.util import combine_opts, ref_to_obj, obj_to_ref

from .util import adict, asdur, addPrefix, makeID, cull, now, ts2dt, dt2ts, resolve
from . import api, broker

log = logging.getLogger(__name__)

#------------------------------------------------------------------------------
def pyramid_scheduler_wrapper(schedulerID, jobID, task):
  log.debug('running job "%s" for scheduler "%s"', jobID, schedulerID)
  Scheduler.registry[schedulerID]._runJob(jobID, task)

#------------------------------------------------------------------------------
def housekeeping(schedulerID):
  log.debug('starting housekeeping for scheduler "%s"', schedulerID)
  Scheduler.registry[schedulerID]._housekeeping()

#------------------------------------------------------------------------------
class Scheduler(object):

  #: the amount of time to add to **now** in order to ensure that a
  #: `date` job that is scheduled for now does not get refused. this
  #: is to work around an APS "peculiarity" that does not allow a job
  #: to be scheduled for now, as in *right* *now*...
  # TODO: this should be re-implemented to auto-convert to an `async`
  #       job instead...
  TIMEPAD = 0.01

  #: if no queue names are specified, `DEFAULT_QUEUE` is the name
  #: given to the one and only queue that will be created.
  DEFAULT_QUEUE = 'default'

  registry = dict()

  #----------------------------------------------------------------------------
  def __init__(self, settings=None, appreg=None, **kw):
    self._listeners = []
    self.id         = None
    self.conf       = adict()
    self.aps        = None
    self.ramstore   = 'internal.transient.' + makeID()
    self.appreg     = appreg
    if settings is not None:
      self.configure(settings, **kw)

  #----------------------------------------------------------------------------
  def configure(self, settings, prefix='scheduler.'):
    self.confdict = conf = combine_opts(settings, prefix)
    self.conf.combined     = asbool(conf.get('combined', True))
    self.conf.housekeeping = asdur(conf.get('housekeeping', '24h'))
    self.conf.hkpostcall   = resolve(conf.get('housekeeping.append', None))
    self.conf.queues       = aslist(conf.get('queues', ''))
    self.conf.grace        = int(conf.get('misfire_grace_time', 1))
    self.conf.delegator    = resolve(conf.get('delegator', None))
    self.id                = conf.get('id', ';'.join(self.conf.queues))
    if len(self.conf.queues) <= 0:
      self.conf.queues = [self.DEFAULT_QUEUE]
    self.broker = broker.Engine(self, conf)
    self.registry[self.id] = self

  #----------------------------------------------------------------------------
  def startProducer(self):
    '''
    Enables this scheduler to "produce" (i.e. create) jobs - must be
    called prior to any calls to the
    `add_(async|date|interval|cron)_job()` methods. Any number of
    schedulers can be started as "producers".
    '''
    pass

  #----------------------------------------------------------------------------
  def startConsumer(self, daemon=True, queues=[]):
    '''
    Enables this scheduler to "consume" (i.e. execute)
    jobs. Currently, this method must be called only **ONCE** for any
    given queue, as it will launch a thread that actually executes the
    jobs. The caller is in charge of ensuring this concurrency
    limitation; it is therefore recommended that `pscheduler` be used
    with a process manager (such as `supervisor
    <http://cr.yp.to/daemontools.html>`_) in conjunction with the
    configuration `scheduler.combined` set to `false`.

    :param daemon:

      Specifies whether or not the thread created for the consumer
      will be set to a daemon thread, i.e. if True (the default), it
      will stop as soon as the main and all non-daemon threads have
      stopped.
    '''
    self.broker.startConsumer(queues)
    log.info('starting APS job execution daemon')
    self.aps = apscheduler.scheduler.Scheduler()
    self.aps.add_jobstore(RAMJobStore(), self.ramstore)
    apsconf = addPrefix(self.confdict, 'apscheduler.')
    apsconf['apscheduler.standalone'] = 'false'
    self.aps.configure(apsconf, daemonic=daemon)
    self.aps.add_interval_job(housekeeping, args=(self.id,),
                              seconds=self.conf.housekeeping, jobstore=self.ramstore)
    self.aps.add_listener(self._apsEvent)
    self.aps.start()

  #----------------------------------------------------------------------------
  def shutdown(self):
    '''
    Shutdown any active producer/consumer threads and/or processes.
    '''
    # todo: shutdown the self.broker and self.aps...
    pass

  #----------------------------------------------------------------------------
  def _apsError(self, event):
    def getJid(event):
      try:    return event.job.id
      except: return None
    log.error('job ID "%s" failed', getJid(event),
              exc_info=event.exception, extra=dict(job=event.job))
    # todo: self._notify?

  #----------------------------------------------------------------------------
  def _apsEvent(self, event):
    if getattr(event, 'exception', None) is not None:
      return self._apsError(event)
    if event.code in (
      apscheduler.events.EVENT_SCHEDULER_START,
      apscheduler.events.EVENT_SCHEDULER_SHUTDOWN,
      apscheduler.events.EVENT_JOBSTORE_ADDED,
      apscheduler.events.EVENT_JOBSTORE_REMOVED,
      #apscheduler.events.EVENT_JOBSTORE_JOB_ADDED,
      #apscheduler.events.EVENT_JOBSTORE_JOB_REMOVED,
      #apscheduler.events.EVENT_JOB_EXECUTED,
      #apscheduler.events.EVENT_JOB_ERROR,
      #apscheduler.events.EVENT_JOB_MISSED,
      ):
      # these can be ignored
      return
    if isinstance(event, apscheduler.events.JobStoreEvent) \
        and event.alias == self.ramstore:
      # this is an "internal" event... squelch it.
      return
    if event.code == apscheduler.events.EVENT_JOBSTORE_JOB_ADDED \
        and hasattr(event, 'job'):
      return self._notify(api.Event(api.Event.JOB_CREATED, job=event.job))
    if event.code == apscheduler.events.EVENT_JOB_EXECUTED:
      return self._notify(api.Event(api.Event.JOB_EXECUTED, job=event.job))
    if event.code == apscheduler.events.EVENT_JOBSTORE_JOB_REMOVED \
        and hasattr(event, 'job'):
      return self._notify(api.Event(api.Event.JOB_REMOVED, job=event.job))
    # todo: any other messages that i should pass through?
    # print 'APS.EVENT:',repr(event)

  #----------------------------------------------------------------------------
  def addListener(self, listener, events=None, args=None, kwargs=None):
    '''
    Adds the callable `listener` to be notified of the events listed
    in `events`, which is a list of constants from
    :class:`pyramid_scheduler.api.Event` that the listener is
    interested in receiving notifications for. Settings `events` to
    None (the default) is equivalent to listening to all
    events. `args` and `kwargs` will be passed to `listener` after the
    first parameter, the :class:`pyramid_scheduler.api.Event` itself.

    Several important notes to be aware of:

    * These events will only be triggered in the process that
      called :meth:`startConsumer`.

    * The listeners will be called from the scheduler thread, so keep
      callback duration to a minimum (this restriction may be lifted
      later by creating a new thread for notification dispatch).

    * The listeners will NOT be called from within a pyramid
      transaction, so if transaction context is needed (i.e. database
      access), create one, e.g.::

        import transaction
        def listener(event, *args, **kw):
          with transaction.manager:
            # ...
    '''
    # todo: mutex?
    self._listeners.append((listener, events, args, kwargs))

  #----------------------------------------------------------------------------
  def removeListener(self, listener):
    '''
    Removes the callable `listener` from receiving any further event
    notifications.
    '''
    # todo: mutex?
    self._listeners = [l for l in self._listeners if l[0] is not listener]

  #----------------------------------------------------------------------------
  def _notify(self, event):
    # todo: mutex?
    for sub in self._listeners:
      if sub[1] is not None and event.type not in sub[1]:
        continue
      try:
        sub[0](event, *sub[2] or [], **sub[3] or {})
      except Exception:
        log.exception('notifying listener %r of event %r failed')

  #----------------------------------------------------------------------------
  def _getdt(self, ts, asStart=False, grace=None):
    if not ts:
      return ts
    if not asStart:
      if isinstance(ts, int) or isinstance(ts, float):
        return ts2dt(ts)
      return ts
    if not ( isinstance(ts, int) or isinstance(ts, float) ):
      ts = dt2ts(ts)
    if ts > ( now() + self.TIMEPAD ):
      return ts2dt(ts)
    if grace is None:
      grace = self.conf.grace
    if ts < ( now() - grace ):
      return ts2dt(ts)
    return ts2dt(now() + self.TIMEPAD)

  #----------------------------------------------------------------------------
  def _handleEvent(self, event):
    if event.type == api.Event.MESSAGE:
      log.debug('received message event: %s', event.message)
      if event.message == 'print-jobs' and self.aps:
        return self.aps.print_jobs()
      return self._notify(event)
    if event.type == api.Event.JOB_CREATED:
      return self._handleEventCreated(event)
    if event.type == api.Event.JOB_CANCELED:
      return self._handleEventCanceled(event)
    log.error('unhandled event: %r', event)

  #----------------------------------------------------------------------------
  def _handleEventCreated(self, event):
    if not self.aps:
      return
    job = event.job
    # APS has this silly bug where if i schedule a job for 'now' it
    # refuses to run it with "ValueError('Not adding job since it
    # would never be run')" -- so the _getdt() calls are to work around
    # that...
    params = job.options
    if 'start_date' in params:
      params.start_date = self._getdt(params.start_date, asStart=True, grace=params.misfire_grace_time)
    if job.type == 'async':
      # todo: make this thread's profile match the ones created by APScheduler...
      thread = threading.Thread(
        target = pyramid_scheduler_wrapper,
        args   = (self.id, job.id, job.task),
      )
      thread.daemon = True
      thread.start()
      return
    if job.type == 'date':
      params.date = self._getdt(params.date, asStart=True, grace=params.misfire_grace_time)
      self.aps.add_date_job(
        pyramid_scheduler_wrapper,
        args=(self.id, job.id, job.task), **params)
      return
    if job.type == 'interval':
      self.aps.add_interval_job(
        pyramid_scheduler_wrapper,
        args=(self.id, job.id, job.task), **params)
      return
    if job.type == 'cron':
      self.aps.add_cron_job(
        pyramid_scheduler_wrapper,
        args=(self.id, job.id, job.task), **params)
      return
    log.error('unknown job scheduling type "%s" in event: %r', job.type, event)

  #----------------------------------------------------------------------------
  def _handleEventCanceled(self, event):
    if not self.aps:
      return
    job = event.job
    if not job.id:
      log.error('received request to cancel unidentified job: %r', job)
      return
    jobs = [j for j in self.aps.get_jobs() if j.id == job.id]
    if not jobs:
      log.warning('received request to cancel unknown job: %r', job)
      return
    for apsjob in jobs:
      self.aps.unschedule_job(apsjob)
      self._notify(api.Event(api.Event.JOB_CANCELED, job=apsjob))

  #----------------------------------------------------------------------------
  @contextmanager
  def _threadContext(self):
    # IMPORTANT: this assumes that APScheduler invokes jobs in a separate
    #            thread per job (as documented)...
    # TODO: this needs confirmation!
    if self.appreg is None:
      yield
    else:
      from pyramid.threadlocal import manager
      reg = dict(manager.get())
      reg['registry'] = self.appreg
      manager.push(reg)
      try:
        yield
      finally:
        manager.pop()

  #----------------------------------------------------------------------------
  def _runJob(self, jobID, task):
    func = ref_to_obj(task.func)
    if self.conf.delegator:
      if self.conf.delegator(func, args=task.args, kwargs=task.kwargs):
        return
    with self._threadContext():
      # TODO: ensure proper transaction handling
      with transaction.manager:
        func(*task.args or [], **task.kwargs or {})

  #----------------------------------------------------------------------------
  def _housekeeping(self):
    with self._threadContext():
      # TODO: do housekeeping:
      #         - delete old kombu messages (can't they be auto-delete???)
      if self.conf.hkpostcall:
        self.conf.hkpostcall()

  #----------------------------------------------------------------------------
  def add_async_job(self, func, args=None, kwargs=None, queue=None, **options):
    '''
    Adds an asynchronous job -- identical to calling :meth:`add_date_job` with
    `date` set to the current time (i.e. ``time.time()``).

    Note that this job will appear in the scheduler's job list
    *asynchronously*, i.e. only after some undefined (but short, if
    everything is working correctly) period of time.

    :Returns:

    str
      The identifier for the job.
    '''
    return self._add(func, args, kwargs, queue, 'async', options)

  #----------------------------------------------------------------------------
  def add_date_job(self, func, date, args=None, kwargs=None, queue=None, **options):
    '''
    Schedules the function `func` to be run once on the specified
    `date`, which can be either a datetime (`datetime.datetime`)
    instance or epoch timestamp (int or float). The function will be
    called with positional parameters specified by `args` and keyword
    parameters `kwargs`, both of which must be pickleable via
    `pickle.dumps()` using the highest protocol available. The
    function `func` must be a static module-level function; it
    therefore must not be an anonymous or lambda function, class or
    instance method.

    You can also specify a `queue` that this job should be scheduled
    on, which must be one of the queues defined in the
    pyramid-scheduler configuration. If not specified, it defaults to
    the first specified queue.

    Any of the additional parameters supported by
    :class:`apscheduler.scheduler.Scheduler` are also supported
    in blind pass-through mode.

    Note that this job will appear in the scheduler's job list
    *asynchronously*, i.e. only after some undefined (but short, if
    everything is working correctly) period of time.

    :Returns:

    str
      The identifier for the job.
    '''
    return self._add(func, args, kwargs, queue, 'date',
                     cull(**adict(date=date).update(options)))

  #----------------------------------------------------------------------------
  def add_interval_job(self, func, start_date=None, args=None, kwargs=None, queue=None,
                       weeks=0, days=0, hours=0, minutes=0, seconds=0,
                       **options):
    '''
    A repeating interval-oriented version of :meth:`add_date_job` --
    see :meth:`apscheduler.scheduler.Scheduler.add_interval_job` for
    additional documentation.

    :Returns:

    str
      The identifier for the job.
    '''
    # todo: this method of resolving weeks/days/hours/minutes to
    #       seconds loses the ability to account for DST, leap-days,
    #       leap-seconds, etc. FIX!
    seconds += ( weeks * 604800 ) + ( days * 86400 ) + ( hours * 3600 ) + ( minutes * 60 )
    return self._add(func, args, kwargs, queue, 'interval',
                     cull(**adict(start_date=start_date, seconds=seconds).update(options)))

  #----------------------------------------------------------------------------
  def add_cron_job(self, func, start_date=None, args=None, kwargs=None, queue=None,
                   year=None, month=None, day=None, week=None, day_of_week=None,
                   hour=None, minute=None, second=None, **options):
    '''
    A cron-like version of :meth:`add_date_job` -- see
    :meth:`apscheduler.scheduler.Scheduler.add_cron_job` for
    additional documentation.

    :Returns:

    str
      The identifier for the job.
    '''
    return self._add(func, args, kwargs, queue, 'cron',
                     cull(**adict(start_date=start_date, year=year, month=month,
                                  day=day, week=week, day_of_week=day_of_week,
                                  hour=hour, minute=minute, second=second).update(options)))

  #----------------------------------------------------------------------------
  def _add(self, func, args, kwargs, queue, ftype, fargs):
    queue = queue or self.conf.queues[0]
    if queue not in self.conf.queues:
      raise api.InvalidQueue(queue)
    event = api.Event(
      api.Event.JOB_CREATED,
      queue = queue,
      job   = adict(
        id      = makeID(),
        type    = ftype,
        options = adict(fargs),
        task    = adict(
          func    = obj_to_ref(func),
          args    = args,
          kwargs  = kwargs,
        )
      )
    )
    self.broker.send(event, queue)
    return event.job.id

  #----------------------------------------------------------------------------
  def cancel_job(self, jobID, queue=None):
    '''
    Cancels the job with the ID `jobID` (as returned by any of the
    `add_(async|date|interval|cron)_job()` methods). If `queue` is not
    specified, it is removed from all queues. Note that the execution
    of this request is *asynchronous*.
    '''
    if queue and queue not in self.conf.queues:
      raise api.InvalidQueue(queue)
    event = api.Event(api.Event.JOB_CANCELED, job=adict(id=jobID))
    self.broker.send(event, queue)


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------
